anvil-python/anvil/commands/upgrades/inter_channel.py

```diff
@@ -245,32 +245,42 @@ def get_plan(self) -> list[BaseStep]:
 
         Return the steps to complete this upgrade.
         """
-        plan = [
-            UpgradeHAProxyCharm(
-                self.client,
-                self.jhelper,
-                self.manifest,
-                self.deployment.infrastructure_model,
-            ),
-            UpgradePostgreSQLCharm(
-                self.client,
-                self.jhelper,
-                self.manifest,
-                self.deployment.infrastructure_model,
-            ),
-            # TODO: Don't allow updating MAAS until upgrade path sorted
-            # UpgradeMAASAgentCharm(
-            #     self.client,
-            #     self.jhelper,
-            #     self.manifest,
-            #     self.deployment.infrastructure_model,
-            # ),
-            # UpgradeMAASRegionCharm(
-            #     self.client,
-            #     self.jhelper,
-            #     self.manifest,
-            #     self.deployment.infrastructure_model,
-            # ),
-            UpgradePlugins(self.deployment, upgrade_release=True),
-        ]
+        plan: list[BaseStep] = []
+        if self.client.cluster.list_nodes_by_role("haproxy"):
+            plan.append(
+                UpgradeHAProxyCharm(
+                    self.client,
+                    self.jhelper,
+                    self.manifest,
+                    self.deployment.infrastructure_model,
+                )
+            )
+        if self.client.cluster.list_nodes_by_role("database"):
```

Comment:
cluster has at least one database, otherwise we cannot have maas. As such this check is redundant and can be omitted.